service/src/main/java/bio/terra/profile/service/spendreporting/azure/AzureSpendReportingService.java

```diff
@@ -0,0 +1,126 @@
+   * @param billingProfile The billing profile associated with the billing project.
+   * @param from Start of the billing period
+   * @param to End of the billing period
+   * @return Spend data
+   */
+  public SpendData getBillingProjectSpendData(
+      BillingProfile billingProfile, OffsetDateTime from, OffsetDateTime to) {
+    String k8sNodeResourceGroup =
+        getK8sNodeResourceGroupName(
+            billingProfile.getRequiredTenantId(),
+            billingProfile.getRequiredSubscriptionId(),
+            billingProfile.getRequiredManagedResourceGroupId());
+
+    List<CompletableFuture<SpendData>> azureCostManagementQueriesFutures =
+        List.of(
+            CompletableFuture.supplyAsync(
+                () ->
+                    querySpendDataForResourceGroup(
+                        billingProfile.getRequiredSubscriptionId(),
+                        billingProfile.getRequiredManagedResourceGroupId(),
+                        from,
+                        to)),
+            CompletableFuture.supplyAsync(
+                () ->
+                    querySpendDataForResourceGroup(
+                        billingProfile.getRequiredSubscriptionId(),
+                        k8sNodeResourceGroup,
+                        from,
+                        to)));
+
+    List<SpendData> spendDataList =
+        azureCostManagementQueriesFutures.stream().map(CompletableFuture::join).toList();
+
+    return new SpendData(
+        spendDataList.stream().flatMap(sp -> sp.getSpendDataItems().stream()).toList(), from, to);
+  }
+
+  private SpendData querySpendDataForResourceGroup(
+      UUID subscriptionId, String resourceGroupName, OffsetDateTime from, OffsetDateTime to) {
+    Response<QueryResult> costQueryResponse =
+        azureCostManagementQuery.resourceGroupCostQueryWithResourceTypeGrouping(
+            subscriptionId, resourceGroupName, from, to);
+    if (isK8sResourceGroup(resourceGroupName)) {
+      return queryResultMapper.mapQueryResult(
+          costQueryResponse.getValue(), SpendCategoryType.COMPUTE);
+    } else {
+      return queryResultMapper.mapQueryResult(costQueryResponse.getValue());
+    }
+  }
+
+  private boolean isK8sResourceGroup(String resourceGroupName) {
+    return resourceGroupName.startsWith(K8S_RESOURCE_GROUP_NAME_PREFIX);
+  }
+
+  /**
+   * Returns the name of the corresponding K8s node resource group.
+   *
+   * @param resourceGroupName Name of the managed resource group
+   * @return Name of the K8s node resource group
+   */
+  private String getK8sNodeResourceGroupName(
+      UUID tenantId, UUID subscriptionId, String resourceGroupName) {
+    ResourceManager resourceManager = crlService.getResourceManager(tenantId, subscriptionId);
+    Optional<GenericResource> k8sResource =
+        resourceManager.genericResources().listByResourceGroup(resourceGroupName).stream()
+            .filter(gr -> gr.resourceType().startsWith(AZURE_KUBERNETES_RESOURCE_TYPE))
+            .findFirst();
```

Comment:
It means there are 2 or more K8s cluster in MRG. We don't expect such situation now. Probably make sense to throw/log exception.

Reply:
I'm on the fence about whether we should throw an exception if there's more than 1, or if we should just return the cost of all of then since it doesn't really feel like it's the spend report's job to verify the MRG configuration (which may change over time). Leaning towards throwing the exception because it saves the need to overly generalize the code at this point.

Reply:
Log and throw.